### ISSUE TITLE
/projects should be watchable

### ISSUE THREAD

**Issue by stefwalter** [openshift/origin#6630]
UIs need to be able to watch when the ProjectList at /projects changes.

There’s no way in the Openshift API to watch projects. Since a project IS A namespace … one would think you could watch namespaces instead and infer the existence of projects. But that’s not possible for folks without cluster admin level privileges. Namespaces are not listable or watchable by other users.

**Comment by stefwalter** [comment]
```
[root@f1 ~]# oc version
oc v1.1-668-g705a7dc
kubernetes v1.1.0-origin-1107-g4c8e6f4
```

**Comment by stefwalter** [comment]
$ oc whoami
scruffy
$ oc proxy &
Starting to serve on 127.0.0.1:8001
$ curl http://localhost:8001/oapi/v1/projects?watch=true
{
  "kind": "ProjectList",
  "apiVersion": "v1",
...
$ curl http://localhost:8001/oapi/v1/namespaces?watch=true
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "User \"scruffy\" cannot list all namespaces in the cluster",
  "reason": "Forbidden",
  "details": {
    "kind": "namespaces"
  },
  "code": 403
}
```

**Comment by liggitt** [comment]
The semantics of watch on an ACL-filtered list are unclear. Would you expect an ADD event for a project when you were granted access to a project (even if the namespace previously existed)? Would you expect a DELETE event for a project when your access was removed (even if the namespace still existed)?

**Comment by stefwalter** [comment]
Yes, you described it well.

A poor mans work around would be to invent another mechanism in /oapi that can be watched ... and fires when any changes, and the caller can retrieve the ACL filtered ProjectList again. Taken as a whole this is even less efficient ...

**Comment by liggitt** [comment]
We'd also want to make sure clients didn't display terrifying messages like "Project X was deleted", when maybe you just had your access removed

**Comment by deads2k** [comment]
If we're ok with slightly weird messages about creation and deletion (you get them based on access, not based on actual creation/deletion), our existing project cache already performs all the correct computations.  We'd wire a listener into it that handles the diff and does channel based notifications.

It's not trivial, but I think its possible.

**Comment by liggitt** [comment]
resourceVersion semantics would be wacky

**Comment by deads2k** [comment]
> resourceVersion semantics would be wacky

It would be simplest not to support it.  You get a watch of project-cache updates from the point in time at which you request it.  ResourceVersions would match actual project resource versions.

**Comment by liggitt** [comment]
fixed in #6828
